### DISCUSSION TITLE
OR-Tools CVRP com várias viagens

### DISCUSSION THREAD

**Discussion by Ezequiel-orloski** [google/or-tools#3245] [Routing (and legacy CP) questions]
I'm using OR-Tools Solver to solve a Multi Trip Enabled VRP. what i need is

 - a depot, the route starts and ends here.
 - unloading locations (different from the deposit)
 
Thus, vehicles must pick up goods from each node until the capacity is filled. Then choose an "unloading location", unload all your weight and continue collecting demand from nodes until a time limit is reached or all goods are collected. Then go back to the depot.

The CVRP Recharge Example seems very close, but in my case, the location chosen for unloading must be the second node visited after the depot, and at the end of the route the vehicles must visit the unloading place before the depot. In other words, a vehicle cannot go to the depot (start, end location) with cargo.

Examplo:
```
0: Depot
1 - 10: Unloading Location
11 - N: Nodes to pick up demand

0 > 12 > 13 > 14 > 15 > 16 > 17 > 0  (current)
0 > (1 - 10)(unload) > 12 > 13 > 14 > (1 - 10)(unload) > 15 > 16 > 17 > (1 - 10)(unload) > 0   (what I need)
```

in this case, I can have several unloading locations, but I can choose only 1, and use it on the entire route, and it must be the second and penultimate in the one to be visited

I'm new to or-tools and I'm trying to figure this out. Any help will be appreciated.

Edit 1 I managed to get the vehicle to pass at the unloading location before returning to the depot, using the following:

```
capacity_dimension = routing.GetDimensionOrDie("Capacity")
for v in range(manager.GetNumberOfVehicles()):
    end = routing.End(v)
    capacity_dimension.SetCumulVarSoftUpperBound(end, 0, 100_000)
```

**Comment by Mizux** [comment]
1. To force `Start -> unload`: Did you try to use routing.NextVar() to limit the range of nodes a vehicle can visit ?
e.g.
```python
unload_list = list(range(1,11))
unload_indices = [manager.NodeToIndex(node) for node in unload_list]

# force next node of each vehicle start to an unload node or the end node
for v in range(manager.GetNumberOfVehicles()):
  start_index = routing.Start(v) # get start index
  end_index = routing.Start(v) # get end index (in case route is empty vehicle need to be allow to do start -> end)
  routing.NextVar(start_index).SetValues(unload_indices + end_index)
```

2. For `unload -> end` I would instead add a constraint that vehicle must be empty when visiting the end node

So  supposing you have a load dimension to track the current load of each vehicle.
```python
for v in range(manager.GetNumberOfVehicles()):
  end_index = routing.End(v) # get vehicle's end index
  # vehicle is allow to visit end node iif empty
  routing.Solver().Add(load_dim.CumulVar(end_index) == 0)
```

3. For fun with mermaid markdown extension:

```mermaid
flowchart LR
S([Depot]) --> U[unload] ---> A[...] ---> A & U
S & U --> E([End])
```

**Reply by Ezequiel-orloski** [reply]
hello, thanks for the help!
I added this to my code:
```
 for (int vehicle = 0; vehicle < manager.getNumberOfVehicles(); vehicle++) {
     unload_indices[unload_indices.length - 1] = routing.end(vehicle);
     routing.nextVar(routing.start(vehicle)).setValue(unload_indices);
     capacityDimension.setCumulVarSoftUpperBound(routing.end(vehicle), 0, 100000);
 }
```
result:
```
0 -> 2 (Reload) -> 26 -> 14 -> 16 -> 19 -> 9 -> 18 -> 12 -> 11 -> 5 -> 1 (Reload) -> 21 -> 25 -> 8 -> 28 -> 10 -> 24 -> 23 -> 13 -> 20 -> 22 -> 3 (Reload) -> 0
```
the output format is very close to what I need, however, the three discharge nodes used must be the same.

expected result:
```
0 -> 2 (Reload) -> 26 -> 14 -> 16 -> 19 -> 9 -> 18 -> 12 -> 11 -> 5 -> 2 (Reload) -> 21 -> 25 -> 8 -> 28 -> 10 -> 24 -> 23 -> 13 -> 20 -> 22 -> 2 (Reload) -> 0
```
thanks.

**Reply by Mizux** [reply]
no you can't, every node can **ONLY** be visited once at most, (that's why you need to duplicate/have several refill nodes).
Feel free in a post process step to rename/remap all refill nodes to the same name.

Now if you have several refill stations let's say `A`, `B` and `C` (and several associated nodes `A_1`, `A_2`, `A_3`, `B_1`,...) and you want only one refill station location along each route.

1. if you know/want to force the refill stations to a vehicle group then you can use `routing.VehicleVar()`
ref: https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/ortools/constraint_solver/routing.h#L1380-L1382

2. If you don't know which stations is associated to each vehicle BUT you just want to only visit the same refill station, then I would:
    * Add a `station_<id>` routing dimension.
    * At each station node increase the associated station dimension
      e.g. node `A_1` will increase the routing dimension `station_a`
    * For each station node, add a constraint that other `stations_<id>` must be zero aka you can't visit a node belongs to `station_b` if you previously visited a `station_x`...
